src/qiboml/operations/differentiation.py

```diff
@@ -1,42 +1,163 @@
-from qiboml.models.encoding import BinaryEncoding
-
-
-class PSR:
-
-    def __init__(self):
-        self.scale_factor = 1.0
-        self.epsilon = 1e-2
+from qiboml.models.encoding import BinaryEncoding, QuantumEncoding
+
+
+@dataclass
+class DifferentiationRule(ABC):
+
+    @abstractmethod
+    def evaluate(
+        self,
+        x: ndarray,
+        encoding: QuantumEncoding,
+        training: Circuit,  # TODO: replace with an abstract TrainableLayer
+        decoding: QuantumDecoding,
+        backend: Backend,
+        *parameters: ndarray,
+    ):
+        """
+        Evaluate the gradient of a quantum model w.r.t variables and parameters,
+        respectively represented by `x` and `parameters`.
+        """
+        pass
+
+
+class PSR(DifferentiationRule):
+    """
+    Compute the gradient of the expectation value of a target observable w.r.t
+    features and parameters contained in a quantum model using the parameter shift
+    rules.
+    """
 
     def evaluate(self, x: ndarray, encoding, training, decoding, backend, *parameters):
         if decoding.output_shape != (1, 1):
             raise_error(
                 NotImplementedError,
                 "Parameter Shift Rule only supports expectation value decoding.",
             )
-        x = encoding(x) + training
-        gradients = []
-        for i, param in enumerate(parameters):
-            tmp_params = backend.cast(parameters, copy=True)
-            tmp_params = PSR.shift_parameter(tmp_params, i, self.epsilon, backend)
-            x.set_parameters(tmp_params)
-            forward = decoding(x)
-            tmp_params = PSR.shift_parameter(tmp_params, i, -2 * self.epsilon, backend)
-            x.set_parameters(tmp_params)
-            backward = decoding(x)
-            gradients.append((forward - backward) * self.scale_factor)
-        return gradients
+        # construct circuit
+        circuit = encoding(x) + training
+
+        # compute first gradient part, wrt data
+        gradient = self.gradient_from_data(
+            data=x,
+            encoding=encoding,
+            circuit=circuit,
+            decoding=decoding,
+            backend=backend,
+        )
+
+        # compute second gradient part, wrt parameters
+        for i in range(len(parameters)):
+            gradient.append(
+                self.one_parameter_shift(
+                    circuit=circuit,
+                    decoding=decoding,
+                    parameters=parameters,
+                    parameter_index=i,
+                    backend=backend,
+                )
+            )
+        return gradient
+
+    def one_parameter_shift(
+        self, circuit, decoding, parameters, parameter_index, backend
+    ):
+        """Compute one derivative of the decoding strategy w.r.t. a target parameter."""
+        gate = circuit.associate_gates_with_parameters()[parameter_index]
+        generator_eigenval = gate.generator_eigenvalue()
```

Comment:
One question, this has to be computed every time, right? There is no way of caching it, because it depends on the value of the parameter, right?